### PR TITLE
add optional wait flag to scaleWorkload function

### PR DESCRIPTION
**Pull request by kale-amruta**
## Change Overview

Currently ScaleWorkload function waits for the workload to be ready for the function to be complete. This PR adds a flag to make the wait optional

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues

- #1290 

## Test Plan

- [x] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

Used the flag in cassandra blueprint, and changed the function `bringupPod` in restore phase and tested these scenarios
```
- func: ScaleWorkload
      name: bringupPod
      args:
        namespace: "{{ .StatefulSet.Namespace }}"
        name: "{{ .StatefulSet.Name }}"
        kind: StatefulSet
        replicas: "{{ .ArtifactsIn.params.KeyValue.replicaCount }}"
``` 

- Without adding flag `waitForReady` , the `bringupPod` function waited for the pods to be in `ready` state since thats the default behavior
- Added `waitForReady: false` , the `bringupPod` function did not wait for the pod to be in `ready` state and started executing the next function
- Added `waitForReady: true` , the `bringupPod` function waited for the pod to be in `ready` state
- Added `waitForReady: "xyz"` , the `bringupPod` function failed to execute because it needed bool value to be passed

### CODE REVIEW

File: pkg/function/scale_workload.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ const (
 	ScaleWorkloadNameArg      = "name"
 	ScaleWorkloadKindArg      = "kind"
 	ScaleWorkloadReplicas     = "replicas"
+	ScaleWorkloadWait         = "waitForReady"
```
Comment (pavannd1) [inline pkg/function/scale_workload.go:38]:
```suggestion
	ScaleWorkloadWaitArg         = "waitForReady"
```

File: pkg/function/scale_workload.go [hunk 2]
```diff
@@ -136,21 +137,25 @@ func getArgs(tp param.TemplateParams, args map[string]interface{}) (namespace, k
...
 	if err != nil {
-		return namespace, kind, name, replicas, err
+		return namespace, kind, name, replicas, waitForReady, err
 	}
 	err = OptArg(args, ScaleWorkloadKindArg, &kind, kind)
 	if err != nil {
-		return namespace, kind, name, replicas, err
+		return namespace, kind, name, replicas, waitForReady, err
+	}
+	err = OptArg(args, ScaleWorkloadKindArg, &waitForReady, waitForReady)
```
Comment (pavannd1) [inline pkg/function/scale_workload.go:156]:
```suggestion
	err = OptArg(args, ScaleWorkloadWaitArg, &waitForReady, waitForReady)
```

File: pkg/function/scale_workload.go (hunk 1 above)
Comment (pavannd1) [inline pkg/function/scale_workload.go:38]:
```suggestion
	ScaleWorkloadReplicas     = "replicas"
	ScaleWorkloadWaitArg      = "waitForReady"
```

File: docs/functions.rst [hunk 3]
```diff
@@ -178,6 +178,12 @@ to stop a database process before restoring files.
    `name`, No, `string`, name of the workload to scale
    `kind`, No, `string`, `deployment` or `statefulset`
    `replicas`, Yes, `int`,  The desired number of replicas
+   `waitForReady`, No, `bool`, Whether to wait for the workload to be ready before executing next steps
+
+.. note::
+   ``waitForReady`` argument is optional and if the argument is not specified
+   in the function, default value is ``true`` which means ``ScaleWorkload``
+   function will always wait for the workload to be in ``Ready`` state by default
```
Comment (viveksinghggits) [inline docs/functions.rst:186]: maybe we can just specify the default value in the above table itself. but I will let others have a look.

Comment (viveksinghggits) [reply docs/functions.rst:186]: in that case we won't need this paragraph.

Comment (ihcsim) [reply docs/functions.rst:186]: Agreed. See CreateCSISnapshot and RestoreCSISnapshot documentation for existing examples.

File: pkg/function/scale_workload.go [hunk 4]
```diff
@@ -92,14 +93,15 @@ func (*scaleWorkloadFunc) Arguments() []string {
...
 	}
 }
 
-func getArgs(tp param.TemplateParams, args map[string]interface{}) (namespace, kind, name string, replicas int32, err error) {
+func getArgs(tp param.TemplateParams, args map[string]interface{}) (namespace, kind, name string, replicas int32, waitForReady bool, err error) {
 	var rep interface{}
 	err = Arg(args, ScaleWorkloadReplicas, &rep)
 	if err != nil {
-		return namespace, kind, name, replicas, err
+		return namespace, kind, name, replicas, waitForReady, err
```
Comment (ihcsim) [inline pkg/function/scale_workload.go:104]: `waitForReady` is `false` here; probably wants to set it to `true` early. Similarly, at line 141.

File: docs/functions.rst (hunk 3 above)
Comment (kale-amruta) [reply docs/functions.rst]: Makes sense, I have updated it.

File: pkg/function/scale_workload.go (hunk 4 above)
Comment (kale-amruta) [reply pkg/function/scale_workload.go:105]: Done

File: pkg/function/scale_workload.go [hunk 5]
```diff
@@ -92,14 +93,16 @@ func (*scaleWorkloadFunc) Arguments() []string {
...
 }
 
-func getArgs(tp param.TemplateParams, args map[string]interface{}) (namespace, kind, name string, replicas int32, err error) {
+func getArgs(tp param.TemplateParams, args map[string]interface{}) (namespace, kind, name string, replicas int32, waitForReady bool, err error) {
 	var rep interface{}
+	waitForReady = true
 	err = Arg(args, ScaleWorkloadReplicas, &rep)
 	if err != nil {
-		return namespace, kind, name, replicas, err
+		return namespace, kind, name, replicas, waitForReady, err
```
Comment (ihcsim) [inline pkg/function/scale_workload.go:105]: I think we can just do a simple `return` without the trailing arguments. Same with line 148, 152, 156, 160 and 162. Sorry, should have pointed that out earlier.

**Review by ihcsim — APPROVED** [review]
LGTM :+1:. Thanks for working on this!

**Pull request merged**